sample-service/src/main/java/uk/ac/ox/ndph/mts/sample_service/security/authorisation/AuthorisationService.java

```diff
@@ -20,24 +28,46 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(AuthorisationService.class);
 
     private final SecurityContextUtil securityContextUtil;
+    private final SiteTreeUtil siteTreeUtil;
 
     private final PractitionerServiceClient practitionerServiceClient;
     private final RoleServiceClient roleServiceClient;
+    private final SiteServiceClient siteServiceClient;
 
     @Autowired
     public AuthorisationService(final SecurityContextUtil securityContextUtil,
+                                final SiteTreeUtil siteTreeUtil,
                                 final PractitionerServiceClient practitionerServiceClient,
-                                final RoleServiceClient roleServiceClient) {
+                                final RoleServiceClient roleServiceClient,
+                                final SiteServiceClient siteServiceClient) {
         this.securityContextUtil = securityContextUtil;
+        this.siteTreeUtil = siteTreeUtil;
         this.practitionerServiceClient = practitionerServiceClient;
         this.roleServiceClient = roleServiceClient;
+        this.siteServiceClient = siteServiceClient;
     }
 
+    public boolean authorise(String requiredPermission, List<Object> sites, String methodName) {
```

Comment:
Presumably this method is the entry point for services wanting to check that the user has access to the required permission at the given sites? In that case accepting an arbitrary method name is a bit dangerous; also the list should be either a list of something that extends an interface e.g. List<HasSite> for somethings that can return a site (or site ID if they just return the ID). Then there is no need for callers to pass the methodName. If you do need to pass an arbitrary 'site ID getter' in here, prefer to use a lambda over method introspection i.e. signature should be:

    public boolean authorise(String requiredPermission, List<? extends Object> sites, Function<Object, String> idGetter) {...

but prefer: 

    public boolean authorise(String requiredPermission, List<? extends HasSiteId> sites) {...}

and 

```
public interface HasSiteId {
    String getSiteId();
}
```

Reply:
I thought about it a lot actually when writing this piece of code, and I also prefer the interface implementation. But that will mean that most of the models in our services will have to implement that. Is it something we are ok with ?

Reply:
The models won't have to implement it directly, the services can always wrap the actual model in a facade that translates to site IDs (which is essentially what you are doing in the service). The typed alternative is to use the lambda like I said above, or a generic version:

`public <T> boolean authorise(String requiredPermission, List<T> sites, Function<T, String> idGetter) {...
`

so here T is the type you are trying to authorise for, and the idGetter is the functional (lambda) to get the ID given a T.